stm/stm.ml

```diff
@@ -1321,26 +1287,71 @@ end = struct (* {{{ *)
   let get_prev_proof ~doc id =
     try
       let np = get_proof ~doc id in
-      match np with
-      | None -> None
-      | Some cp ->
-        let did = ref id in
-        let rv = ref np in
-        let done_ = ref false in
-        while not !done_ do
-          did := fold_until back_tactic 1 !did;
-          rv := get_proof ~doc !did;
-          done_ := match !rv with
-            | Some rv -> not (Goal.Set.equal (Proof.all_goals rv) (Proof.all_goals cp))
-            | None -> true
-        done;
-        !rv
+      if np = None then
+        (* handle "Show" command, which passes (last valid id)+1 *)
+        get_proof ~doc (fold_until back_tactic 1 id)
+      else
+        match np with
+        | None -> None
+        | Some cp ->
+          let did = ref id in
+          let rv = ref np in
+          let done_ = ref false in
+          while not !done_ do
+            did := fold_until back_tactic 1 !did;
+            rv := get_proof ~doc !did;
+            done_ := match !rv with
+              | Some rv -> not (Goal.Set.equal (Proof.all_goals rv) (Proof.all_goals cp))
+              | None -> true
+          done;
+          !rv
     with Not_found | PG_compat.NoCurrentProof -> None
 
 end (* }}} *)
 
 let get_prev_proof = Backtrack.get_prev_proof
 
+(* Wrapper for Vernacentries.interp to set the feedback id *)
+(* It is currently called 19 times, this number should be certainly
+   reduced... *)
+let stm_vernac_interp ?proof ?route id st { verbose; loc; expr } : Vernacstate.t =
+  (* The Stm will gain the capability to interpret commmads affecting
+     the whole document state, such as backtrack, etc... so we start
+     to design the stm command interpreter now *)
+  set_id_for_feedback ?route dummy_doc id;
+  Aux_file.record_in_aux_set_at ?loc ();
+  (* We need to check if a command should be filtered from
+   * vernac_entries, as it cannot handle it. This should go away in
+   * future refactorings.
+  *)
+  let is_filtered_command = function
+    | VernacResetName _ | VernacResetInitial | VernacBack _
+    | VernacBackTo _ | VernacRestart | VernacUndo _ | VernacUndoTo _
+    | VernacAbortAll | VernacAbort _ -> true
+    | _ -> false
+  in
+  let aux_interp st expr =
+    (* XXX unsupported attributes *)
+    let cmd = Vernacprop.under_control expr in
+    if is_filtered_command cmd then
+      (stm_pperr_endline Pp.(fun () -> str "ignoring " ++ Ppvernac.pr_vernac expr); st)
+    else
+      match cmd with
+      | VernacShow ShowScript -> ShowScript.show_script (); st (* XX we are ignoring control here *)
+      | _ ->
+        stm_pperr_endline Pp.(fun () -> str "interpreting " ++ Ppvernac.pr_vernac expr);
+        let doc = get_doc 0 in
+        let dproof = get_prev_proof ~doc (get_current_state ~doc) in
```

Comment:
I don't think this line is correct, for example if you repair a broken proof `get_current_state` is not returning the state you want.

Reply:
The use case is that a `Show` should display output identical to the last tactic, including the diff highlights shown in the last tactic.

Complicating this is that for proof view-only changes such as `swap`, we want to diff between the current proof state and the proof state before the last tactic (https://github.com/coq/coq/pull/8247/commits/a828bcedb8ad60c5b1f4cf71f92f24f2c1197ecb).  My current code doesn't handle that the same way for `Show` yet.